art.go

```diff
@@ -20,79 +24,131 @@ package art
 
 // TODO: IPv4 + IPv6 constructors
 
-func baseIndex(width int, addr uint64, prefixLen int) uint64 {
-	return (addr >> uint64(width-prefixLen)) + (1 << uint64(prefixLen))
+func baseIndex(width int, addr uint32, prefixLen int) uint32 {
+	return (addr >> uint32(width-prefixLen)) | (1 << uint32(prefixLen))
 }
 
-func fringeIndex(width int, addr uint64) uint64 {
+func fringeIndex(width int, addr uint32) uint32 {
 	return baseIndex(width, addr, width)
 }
 
+func getBits(byteOffset, bits int, src []byte) uint32 {
+	if byteOffset > len(src) {
+		return 0
+	}
+	var v uint32
+	src = src[:(len(src) - byteOffset)]
+	switch len(src) {
+	case 0:
+		return 0
+	case 1:
+		v = uint32(src[0])
+	case 2:
+		v = uint32(src[1]) + (uint32(src[0]) << 8)
+	case 3:
+		v = uint32(src[2]) +
+			(uint32(src[1]) << 8) +
+			(uint32(src[0]) << 16)
+	default:
+		v = uint32(src[len(src)-4]) +
```

Comment:
the compiler recognizes & inlines https://pkg.go.dev/encoding/binary#ByteOrder into one instruction without a call if you just want to write `binary.BigEndian.Uint32`: 

https://cs.opensource.google/go/go/+/refs/tags/go1.16.6:src/encoding/binary/binary.go;l=111